Add --max-depth option to control diagram complexity

Julfried opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [ ] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [ ] Write comprehensive commit messages and/or a good description of what the PR does.
- [ ] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description
This PR adds a new `--max-depth` option to pyreverse that allows basic controlling of the visualization depth of packages and classes in the generated diagrams. The depth is simply calculated by counting dots in qualified names, where depth 0 represents top-level packages/classes. 

For packages, the full path depth is considered. For classes, the depth is calculated based on their containing module to ensure meaningful class diagrams while respecting the depth limit.

Closes #9233

Julfried commented:
Quick update: I've completed the implementation and added tests for the --max-depth option. I believe the feature is working as intended and is ready for review. I’m currently fighting some CI issues, but they seem unrelated to my changes. Any help or advice would be much appreciated!

Pierre-Sassoulas reviewed (dismissed):
LGTM ! Don't worry about the 3.13 CI fail, this is on main too we can rebase once it's fixed. I'll also wait for a review from Andrea before merging.

Julfried commented:
Thanks! Appreciate your review and the clarification about the CI issue. I`ll keep an eye on any updates and am happy to rebase if needed.

Pierre-Sassoulas commented:
Thank you for rebasing, the mypy error is coming from main the other errors seems genuine. (Spelling is checked only in ci not in local pre-commit)

Julfried commented:
Thanks for pointing out the mypy error, I`ll keep an eye on it if it gets fixed in main. How about disabling spelling check for these test cases because "-" is not allowed in Python naming anyway?

Pierre-Sassoulas commented on tests/pyreverse/test_writer.py:
```diff
@@ -257,3 +257,84 @@ def test_package_name_with_slash(default_config: PyreverseConfig) -> None:
+    """Test that nodes are filtered correctly when depth limit is set.
+
+    Depth counting is zero-based, determined by number of dots in path:
+    - pkg                   -> depth 0 (0 dots)
+    - pkg.subpkg           -> depth 1 (1 dot)
+    - pkg.subpkg.module    -> depth 2 (2 dots)
+    - pkg.subpkg.module.submodule -> depth 3 (3 dots)
```
```suggestion
    - 'pkg'                  -> depth 0 (0 dots)
    - 'pkg.subpkg'           -> depth 1 (1 dot)
    - 'pkg.subpkg.module'    -> depth 2 (2 dots)
    - 'pkg.subpkg.module.submodule' -> depth 3 (3 dots)
```
Somthings's very wrong with the way the github action display pylint message but I guess this fixes the spelling

Pierre-Sassoulas commented on tests/pyreverse/test_writer.py:
```diff
@@ -257,3 +257,84 @@ def test_package_name_with_slash(default_config: PyreverseConfig) -> None:
+    """Test class visibility based on their containing module depth.
+
+     Classes are filtered based on their containing module's depth:
+    - MyClass -> depth 0 (no module)
+    - pkg.MyClass -> depth 0 (module has no dots)
+    - pkg.subpkg.MyClass -> depth 1 (module has 1 dot)
+    - pkg.subpkg.mod.MyClass -> depth 2 (module has 2 dots)
```
```suggestion
    - 'MyClass' -> depth 0 (no module)
    - 'pkg.MyClass' -> depth 0 (module has no dots)
    - 'pkg.subpkg.MyClass' -> depth 1 (module has 1 dot)
    - 'pkg.subpkg.mod.MyClass' -> depth 2 (module has 2 dots)
```

Julfried commented:
It seems that the spelling checks still fail, because it checks for words and ignores the quotes. How about adding these testnames to `.pylintrc` spelling dictionary?

```diff
[SPELLING]
# List of comma separated words that should not be checked.
spelling-ignore-words=subpkg,MyClass
```

Julfried commented:
When working on test coverage I noticed tests/data/__init__.py is empty, which impacts pyreverse's ability to detect complete package relationships. This effects validation of the --max-depth functionality since module relationships aren't fully established and the resulting diagrams do not represent the whole structure.

@Pierre-Sassoulas do you maybe remember if this was an intentional design choice, or did this just grow over time?

Note: Addressing this would require modifying all affected test files to properly reflect package relationships. Given the scope, I think handling such changes in a separate PR would be better, if we decide to do this!

Pierre-Sassoulas commented:
I'm not sure. Pylint had a lot of issues with namespace package and missing `__init__.py` in the past, is it the issue ? If you meant 'empty' and not 'missing' I'm not sure that an empty init.py should influence the result ?

Julfried commented:
Thanks for checking! You're right that it's about an empty (not missing) __init__.py. The difference is how many relationships pyreverse detects:
- Empty __init__.py: Shows only direct module dependencies [empty init file](https://is.gd/tOkhU8)
- __init__.py with imports: Shows both module dependencies and package-level relationships [imports in init file](https://is.gd/dZ7UCH) 

I think this effects depth filtering coverage since we're not testing package-level relationship filtering. Although I want to add, that I am not a testing expert.

Here is what I added to __init__.py:
```python
from .clientmodule_test import Ancestor, Specialization
from .suppliermodule_test import Interface, DoNothing, DoNothing2
from .nullable_pattern import NullablePatterns
from .property_pattern import PropertyPatterns
```

Julfried commented:
Thanks for your earlier guidance, Pierre!

I now rebased to fix the mypy error. Also, I made some mistakes when setting up the tests, which resulted in pyreverse generating wrong diagrams. Now I fixed it by using a proper test setup with fixtures and I test multiple depths, which improved test coverage. The issue with pyreverse not properly detecting all relationships when __init__.py still remains. I believe this is why coverage for package-level relationships is still incomplete. 

In my understanding, an empty __init__.py file limits the ability to fully establish package-level relationships, as it doesn’t define imports or explicitly link modules in the package. This might explain why pyreverse struggles with package-level filtering in such cases. However, I’m not entirely sure if this is the root cause or if there’s another factor at play.

Let me know what you think🙂

Julfried commented:
Thanks for your earlier guidance, Pierre!

I now rebased to fix the mypy error and addressed the spelling issues. Also, I made some mistakes when setting up the tests, which resulted in pyreverse generating wrong diagrams. Now I fixed it by using a proper test setup with fixtures and I test multiple depths, which improved test coverage. The issue with pyreverse not properly detecting all relationships when init.py still remains. I believe this is why coverage for package-level relationships is still incomplete.

In my understanding, an empty init.py file limits the ability to fully establish package-level relationships, as it doesn’t define imports or explicitly link modules in the package. This might explain why pyreverse struggles with package-level filtering in such cases. However, I’m not entirely sure if this is the root cause or if there’s another factor at play.

I’m happy to explore this further and improve the tests in a follow-up PR if needed. Let me know what you think🙂

Pierre-Sassoulas commented on pylint/pyreverse/writer.py:
```diff
@@ -129,6 +179,12 @@ def write_classes(self, diagram: ClassDiagram) -> None:
             )
         # inheritance links
         for rel in diagram.get_relationships("specialization"):
```
```suggestion
        for rel in diagram.get_relationships("specialization", depth=self.max_depth):
```
Wonder if it's possible of filtering on depth directly in ``get_relationships`` instead. Seems like it could reduce duplication of ``.should_show_node`` checks when looping on it

Julfried replied:
Thanks for the suggestion! I would agree, that handling everything regarding filtering in one place would benefit maintainability. However I think we still need node filtering when emitting nodes, to determine which ones to show in the final diagram. Not sure about splitting this up into 2 files either. What do you think?

Julfried replied:
Actually, I feel like we could simplify this, by introducing `get_nodes()` alongside `get_relationships()`. Then the writer would look something like this.
```python
def write_packages(self, diagram: PackageDiagram) -> None:
    for module in diagram.get_nodes(depth=self.max_depth):
        # Emit node logic
    for rel in diagram.get_relationships(depth=self.max_depth):
        # Emit edge logic
```
That way all the filtering logic would be handled consistently inside the diagram classes and the printer would only need to focus on rendering the final components. What do you think?

Pierre-Sassoulas replied:
Looks great !

Julfried replied:
Since this is a big change, I am not sure how to approach this. Are you ok with doing this in this PR?

Pierre-Sassoulas reviewed (commented):
Sorry for the delay. Changes seems sensible, I have a possible refactor to suggest. (I'd like a review from Andras if possible though, no pressure @DudeNr33 ;) )

Julfried commented:
No worries, Currently I am also very packed. Will try to take a look at it later this week :)

DudeNr33 commented on pylint/pyreverse/main.py:
```diff
@@ -192,6 +192,22 @@
+            "default": None,
+            "metavar": "<depth>",
+            "type": "int",
+            "group": OPTIONS_GROUPS["DISPLAY"],
```
```suggestion
            "group": OPTIONS_GROUPS["FILTERING"],
```
I think the new option fits the FILTERING group better, as it is similar to e.g. `--show-associated`. If you agree, can you move it up near the other filtering options, as the options are also ordered in the source code accordingly?

DudeNr33 reviewed (changes requested):
Sorry for the delay, I am not very active on GitHub at the moment. 

Thank you for the PR - this is actually a feature I saw in a similar tool and always thought "would be really helpful to have this in pyreverse"!

Regarding the changes (and following the discussion you already had with Pierre), I feel like the logic is better suited in `diadefslib.py`, and could be handled similarly to `--show_associated`. 
But like Pierre I think handling this in a follow-up PR makes sense, so that this one does not get too big.

Besides that I only have on remark regarding the option group. 👍

Julfried commented:
Thanks, appreciate your review. Both of your suggestions make sense - especially the architectural one about moving depth filtering to diadefslib.py alongside other filters like --show_associated. I've moved the --max-depth option to the Filtering and Scope group to better align with similar options.  I am happy to do a refactor in a follow up PR :)

DudeNr33 reviewed (approved):
Thanks! Now I guess only the failing codecov check in the pipeline prevents this from being merged. 
Do you need help with that?

Julfried commented:
Thanks for checking! As I said here:
> Thanks for your earlier guidance, Pierre! I hope this is PR is not getting to large.
> 
> I now rebased to fix the mypy error and addressed the spelling issues. Also, I made some mistakes when setting up the tests, which resulted in pyreverse generating wrong diagrams. Now I use a proper test setup with fixtures and test multiple depths, to improve test coverage. The issue with pyreverse not properly detecting all relationships when init.py is empty still remains. I believe this is why coverage for package-level relationships is still incomplete.
> 
> In my understanding, an empty init.py file limits the ability to fully establish package-level relationships, as it doesn’t define imports or explicitly link modules in the package. This might explain why pyreverse struggles with package-level filtering in such cases. However, I’m not entirely sure if this is the root cause or if there’s another factor at play.
> 
> I’m happy to explore this further and improve the tests in a follow-up PR if needed. Let me know what you think🙂

I believe this is due to empty `__init__.py` files in `tests/data/`. Appreciate your help, if you have a better understand of what is actually happening.

Julfried commented:
Hey @Pierre-Sassoulas , I believe I am not able to resolve codecov, without changing the structure of the testfiles. How about merging this and doing the proposed refactor in a follow-up PR?

Pierre-Sassoulas reviewed (approved):
Let's do that !
